feat(Beaker): Support distro variant configuration

Tiboris opened the pull request:
Before this patch the distro variant could not be configured
and was hardcoded in the source code of the mrack project.
With this patch we support the old way and also a new way
of specifying the distro variants in the provisioning-config.

Updated the example provisioning-config.yaml file with latest
beaker feature examples.

Signed-off-by: Tibor Dudlák <tdudlak@redhat.com>

dav-pascual reviewed (approved):
Changes acknolegded, thanks @Tiboris

Tiboris commented on src/mrack/transformers/beaker.py:
```diff
@@ -45,26 +45,36 @@ async def init_provider(self):
+        distro_variants = self.config.get("distro_variants")
+
+        if distro_variants:
+            variant = distro_variants.get(
+                required_distro, distro_variants.get("default")
+            )
+        # keep this elif for backward compatibility wit mrack <= 1.2.0
         elif re.match(r"(rhel-[8|9])", host["os"]):
             variant = "BaseOS"
         else:  # Default to Server for RHEL7 and Fedora systems
             variant = "Server"
-        return variant
+
+        if "beaker_variant" in host:
+            variant = host["beaker_variant"]
```
Will do

Tiboris commented:
I have
- updated mocked data
- added mocked BeakerTransformer class
- added test for the distro and variant
- fixed the distro not to point to None when distro is not specified in provisioning config
-

miskopo commented on src/mrack/data/provisioning-config.yaml:
```diff
@@ -95,6 +95,17 @@ beaker:  # beaker provider specific values
     distros:
         # list of beaker distros key is used in metadata distro is used in generated jobxml
         fedora-32: Fedora-32%
+        c9s: Centos-Stream-9%
```
I reckon it should be `CentOS-Stream-9%` as on line `107`

Tiboris replied:
I will fix that

Tiboris replied:
Fixed

miskopo commented on tests/unit/test_beaker_transformer.py:
```diff
@@ -0,0 +1,87 @@
+    fedora = {
+        "name": f"fedora.{domain_name}",
+        "role": "client",
+        "group": "client",
```
just a nitpick, but for consistency, this should be a `ipaclient` group

Tiboris replied:
![come on](https://media.giphy.com/media/MwOuiiTfWfWgM/giphy.gif)

miskopo commented:
The added tests are skipped as seen [here](https://dev.azure.com/neoave/mrack/_build/results?buildId=582&view=ms.vss-test-web.build-test-results-tab&runId=1020&resultId=100034&paneView=debug)

```
('/opt/hostedtoolcache/Python/3.9.10/x64/lib/python3.9/site-packages/_pytest/python.py', 182, 'Skipped: async def function and no async plugin installed (see warnings)')
```

Based on the [warnings](ttps://dev.azure.com/neoave/mrack/_build/results?buildId=582&view=results), `asyncio` plugin is missing. I don't see `pytest-asyncio` listed in the `requirements.txt`, which may be the core of a problem

Tiboris commented:
@miskopo we have the test-requirements file for the testing requirements with the pytest-asyncio>=0.14.0 pinned.

miskopo commented:
In that case, either add `--asyncio-mode=auto` to the `pytest` invocation, which will use `asyncio` on all `async` functions, or, use `--asyncio-mode=strict` and add decorator `@pytest.mark.asyncio` to [test on line 82](https://github.com/neoave/mrack/pull/168/files#diff-79e273e4a7507fde6ea450405d641ab119fd936eebb63b5623cbd63c1bf20847R82)

Tiboris commented:
I seee.... 
```
SKIPPED [3] ../../../../../opt/hostedtoolcache/Python/3.9.10/x64/lib/python3.9/site-packages/_pytest/python.py:182: async def function and no async plugin installed (see warnings)
```
in https://dev.azure.com/neoave/mrack/_build/results?buildId=582&view=logs&j=27008f10-b491-547d-a654-20cfee07d03e&t=73113a39-43fd-502e-b4fb-b68667aceda6 

So skips are not treated as failures in out pipelines.

Tiboris commented:
@miskopo I have added @pytest.mark.asyncio to the test

Tiboris commented:
There were still warning:
`##[warning]The 'asyncio_mode' default value will change to 'strict' in future, please explicitly use 'asyncio_mode=strict' or 'asyncio_mode=auto' in pytest configuration file.` 
So i added another commit

Tiboris commented:
@miskopo even with latest commit there is still 
```
=============================== warnings summary ===============================
../../../../../opt/hostedtoolcache/Python/3.9.10/x64/lib/python3.9/site-packages/pytest_asyncio/plugin.py:191
  /opt/hostedtoolcache/Python/3.9.10/x64/lib/python3.9/site-packages/pytest_asyncio/plugin.py:191: DeprecationWarning: The 'asyncio_mode' default value will change to 'strict' in future, please explicitly use 'asyncio_mode=strict' or 'asyncio_mode=auto' in pytest configuration file.
    config.issue_config_time_warning(LEGACY_MODE, stacklevel=2)
```
any tips how to get rid of it?

miskopo commented:
yes, you can get rid of it by stating in `pytest.ini`

```
[pytest]
asyncio_mode = strict
```

Tiboris commented:
I have re-pushed with the pytest.ini change.

miskopo commented:
I've checked the run results available [here](https://dev.azure.com/neoave/mrack/_build/results?buildId=585&view=ms.vss-test-web.build-test-results-tab&runId=1026&resultId=100032&paneView=debug) and everything seems to be in order. Therefore you have my `ACK` for the test part.

miskopo reviewed (approved):
I approve the testing related changes.

dav-pascual reviewed (approved):
good for me. thanks @Tiboris and @miskopo for helping with tests

Tiboris commented on src/mrack/transformers/beaker.py:
```diff
@@ -45,26 +45,37 @@ async def init_provider(self):
+                required_distro, distro_variants.get("default")
+            )
+        # keep this elif for backward compatibility wit mrack <= 1.2.0
         elif re.match(r"(rhel-[8|9])", host["os"]):
```
With the `distro_variants` present in config there will be no fallback to the old logic and we will only use `distro_variants` to lookup the variant. Should I add another test to remove this record and test the value?

pvoborni replied:
Yes, it would be better given that we keep the old logic present.

Tiboris replied:
https://github.com/neoave/mrack/pull/168/commits/90444b0f76d962a6edd29f2361817260a10465d2
